cloudmarker/test/test_esstore.py

```diff
@@ -0,0 +1,73 @@
+"""Tests for EsStore plugin."""
+
+import unittest
+from unittest import mock
+
+from elasticsearch import ElasticsearchException
+
+from cloudmarker.stores import esstore
+
+
+class EsStoreTest(unittest.TestCase):
+    """Tests for EsStore plugin."""
+
+    @mock.patch('cloudmarker.stores.esstore.Elasticsearch.bulk')
+    def test_happy_flow(self, mock_bulk):
+        mock_record = {'record_type': 'firewall_rule'}
+
+        es_store = esstore.EsStore()
+
+        mock_bulk_response = {
+            'errors': '',
+            'items': [
+                {'index': {'status': 200}}
+            ]
+        }
+
+        mock_bulk.return_value = mock_bulk_response
+
+        es_store.write(mock_record)
+        es_store.done()
+
+        mock_bulk.assert_called_once()
+        # In case of successful push the buffer should reset.
+        self.assertEqual(es_store._buffer, '')
+
+    @mock.patch('cloudmarker.stores.esstore.Elasticsearch.bulk')
+    def test_exception_flow(self, mock_bulk):
+        mock_record = {'record_type': 'firewall_rule'}
+
+        es_store = esstore.EsStore()
+
+        mock_bulk.side_effect = ElasticsearchException()
+
+        es_store.write(mock_record)
+        es_store.done()
+
+        # In case of any expception the buffer should not reset.
+        self.assertNotEqual(es_store._buffer, '')
```

Comment:
Not sure if just checking if buffer is empty would be enough to ensure an exception was raised. I'd rather check if the error was logged. Maybe by mocking the logger itself like below.

```
@mock.patch('cloudmarker.stores.esstore._log')
    def test_exception_flow(self, mock_logger, mock_bulk):
```

and then asserting the logs

```
mock_logger.error.assert_called_with('Bulk Index Error: %s: %s', 'ElasticsearchException', mock_bulk.side_effect)
```

Reply:
I agree with @s-nirali. I would suggest first getting rid of asserting on `_buffer` because it is an internal attribute of the class. That's sufficient for this pull request. The fact that the class does not crash due to an `ElasticsearchException` is a valuable test in itself.

Then, if you want to go ahead and assert the logging in this pull request or in a future one, then I would suggest something like this so that the test method is not tightly coupled with the exact error message being logged:

```
self.assertEqual(len(mock_logger.error.mock_calls), 1)
```

Reply:
@susam , is it okay to mock logger since `_log` is also an internal module attribute?

Reply:
@s-nirali You are right. We should not patch `_log` because it is an internal attribute. If we want to test it the right way, then we would have to expose a public function, say, `setLogger()` in the module that would set its internal `_log` variable to a new logger and then use this function to set the logger to a mock. I think this is an overkill just for the purpose of testing. Another option is to not care that `_log` is internal and patch it anyway. Yet another option is to not bother about testing logger calls.

Reply:
Then what is the best assertion that falls under this? I am still confused about what should I choose the assert statement to be?
I am with @susam  on this about not using the internal state of the class attributes.